### ISSUE TITLE
Migrate to Prosemirror

### ISSUE THREAD

**Issue by tommoor** [outline/rich-markdown-editor#134]
We're investigating moving a future version of this project to Prosemirror for a couple of reasons:

- Out of the box collaborative editing support
- More stable API. Slate's API completely changes several times a year as it is still being refined, and it's caused ([and continues to cause](https://github.com/ianstormtaylor/slate/blob/master/packages/slate/Changelog.md#0500--november-27-2019)) months of extra work.
- More performant
- Improved support for soft keyboards (non-latin languages / mobile)

Ideally, if this comes to fruition the external API would remain largely the same, with the exception of losing support for custom Slate plugins.

**Comment by tommoor** [comment]
Possible base: https://github.com/hubgit/react-prosemirror

**Comment by audiolion** [comment]
Decided to use your editor because it was offloading the management of slate's numerous breaking API changes to this lib to deal with. I feel like much of slate's community of plugins has not kept up with the numerous breaking changes, it is an unfortunate situation. I would not be opposed to a prosemirror swap, and collaborative editing is a huge win. The largest downside of prosemirror is that the API is harder to grok (in my opinion), and takes more time to really wrap your head around. Slate and its plugin system has been very simple to get up and running with.

**Comment by tommoor** [comment]
The spike has gone well, I'm moving forward with migrating the entire editor to Prosemirror – I'd say it's about 60% done already, most of the basics are in place with just some of the more complicated pieces left to do like image uploads, todo lists, and tables.

**Comment by tommoor** [comment]
Fixes #147 
Fixes 126

**Comment by audiolion** [comment]
> Currently I'm unsure if to just make it a new opensource project and archive this one or to make it a major version upgrade. Open to opinions.

I guess it depends on the number of breaking API changes? This repo already has traction from stars so it might be nice to keep that, but if the repo is entirely different and the public API has changed maybe it warrants a new project.

**Comment by tommoor** [comment]
I'll be keeping the main props the same where possible, so unless you're using custom slate plugins it should end up largely a drop-in replacement 😄

**Comment by lordkada** [comment]
I'd keep this repo and make a major version change... according to the number of stars it is quite popular :-) my 2 cents

**Comment by bartzy** [comment]
@tommoor Seems like really nice work so far!
I'm interested - Have you gone with react-prosemirror or something else, or decided to roll your own?
Are you using or plan to use React at all to render nodes in ProseMirror, or just plain JS?
What has been your experience so far with ProseMirror as opposed to Slate?

**Comment by tommoor** [comment]
> Have you gone with react-prosemirror or something else

It's custom, but based on the architecture of [tiptap](https://github.com/scrumpy/tiptap).

> plan to use React at all to render nodes in ProseMirror, or just plain JS?

It's a mixture, where possible plain JS is used for performance.

> What has been your experience so far with ProseMirror as opposed to Slate?

Similar from a learning curve POV, but far more performant for the resulting editor.

**Comment by bartzy** [comment]
Thanks! 
What's the general idea(s) behind the architecture?

Regarding node views with react - when isn't it possible, or not preferable to use React in them?

**Comment by wfpaisa** [comment]
hello @tommoor it would be great to have the possibility to enter formats in a similar way to [Marktext](https://marktext.app/) and [Notion](https://www.notion.so/).

Outline is a very powerful tool, thanks for having this opensource project. 

![DeepinScreenshot_select-area_20200416215901](https://user-images.githubusercontent.com/1735752/79527809-1ce76880-802e-11ea-97ed-2f608b7f43ec.png)
![DeepinScreenshot_select-area_20200416215935](https://user-images.githubusercontent.com/1735752/79527817-1fe25900-802e-11ea-9ce3-05df20b531b5.png)
![DeepinScreenshot_select-area_20200416215949](https://user-images.githubusercontent.com/1735752/79527818-1fe25900-802e-11ea-963a-3daade169db4.png)
![DeepinScreenshot_select-area_20200416220121](https://user-images.githubusercontent.com/1735752/79527820-207aef80-802e-11ea-9345-8fad2b48c864.png)

**Comment by tommoor** [comment]
Yep, the "/" version is already built in the migration branch 👍

**Comment by trompx** [comment]
Hey @tommoor,

I tried to extend the editor by adding custom Nodes, based on the Image.tsx file, even changing the regex IMAGE_INPUT_REGEX but no matter what I do, when I add the new custom type, it is first added with the right Node type.
But after reloading the editor (loads data from localStorage then parses it), it always returns the "image" Node type. I don't think it comes from the IMAGE_INPUT_REGEX as it is used while typing.

**Comment by tommoor** [comment]
@trompx `toMarkdown` and `parseMarkdown` need changing too otherwise it's going to serialize as an image in the markdown.

**Comment by trompx** [comment]
Thanks for the input @tommoor.

The custom node will render approximatly the same as an image so it made sense to have it in markdown in the same `![alt](src)` format. But you mean you cannot have another `![alt](src_with_specific_pattern)` with a pattern to differentiate the node through a regex?

I tried to found what part of the code was responsible of matching the markdown block with a node type, I saw some logic in ExtensionManager but I guess all the work is done in the prosemirror MardownParser.
I tried to set a highest priority in my custom node schema to no avail.
I tried different toMarkdown like:
```
INPUT_REGEX = `/!!!pattern:(\S+)(?:)/`;

toMarkdown(state, node) {
    state.write("!!!pattern:" + state.esc(node.attrs.src));
}

parseMarkdown() {
  return {
    node: "custom",
    getAttrs: token => ({
      src: token.attrGet("src"),
    }),
  };
}
```
but the node type is text in that case.

I have hard time knowing exactly what code is used to determine the node type of a markdown block and if this is what parseMarkdown is doing? And if a regex is used at all.

I managed to get some info about the token used inside the parseMarkdown function, by logging the tokens inside embed.ts file (markdownEmbeds function) but Token attrs are all null.
But I have in fact a token.children[0].content when adding images.
